### ISSUE TITLE
$base_schema_def['foreign keys']  is not returning an array when adding a new Chado Organism Reference

### ISSUE THREAD

**Issue by srobb1** [tripal/tripal#1432]
# BUG/ERROR report

## System information
Mac with Docker Desktop 4.17.0 (99724) 

### Tripal Version:
4.x
tripalproject/tripaldocker:latest
289cbfd0aa8c

### Issue description
When I try to add a new field, Chado Organism Reference, to a gene content type I get this error.
```
Message | TypeError: array_keys(): Argument #1 ($array) must be of type array, null given in array_keys()
(line 74 of /var/www/drupal9/web/modules/contrib/tripal/tripal_chado/src/Plugin/Field/FieldType/obi__organism.php)
#0 /var/www/drupal9/web/modules/contrib/tripal/tripal_chado/src/Plugin/Field/FieldType/obi__organism.php(74): array_keys(NULL)
```

this is line 74 in /tripal/tripal_chado/src/Plugin/Field/FieldType/obi__organism.php
```$base_fk_col = array_keys($base_schema_def['foreign keys']['synonym']['columns'])[0];```

the `$base_schema_def['foreign keys']` is not returning an array, it is returning nothing.

#### Steps to reproduce
Home  -> Administration -> Structure -> Tripal Content Type -> Edit Gene -> Manage Fields -> Add Field -> Chado Organism Reference

**Comment by dsenalik** [comment]
Lacey was asking me yesterday why I had this block in my field

```
    // If we don't have a base table then we're not ready to specify the
    // properties for this field.
    if (!$base_table) {
      $record_id_term = 'SIO:000729';
      return [
        new ChadoIntStoragePropertyType($entity_type_id, self::$id, 'record_id', $record_id_term, [
          'action' => 'store_id',
          'drupal_store' => TRUE,
        ])
      ];
    }

```

and it is because I copied it from the `chado_linker__prop` field, because when adding my field I had a similar error because there was no base table defined.

If I add a similar block to this field, it progresses beyond this error and allows me to add this field.

i.e. after this line
```
   $base_table = $settings['base_table'];
```
I added this
```
    // If we don't have a base table then we're not ready to specify the
    // properties for this field.
    if (!$base_table) {
      $record_id_term = 'OBI:0100026';
      return [
        new ChadoIntStoragePropertyType($entity_type_id, self::$id, 'record_id', $record_id_term, [
          'action' => 'store_id',
          'drupal_store' => TRUE,
        ])
      ];
    }
```

**Comment by srobb1** [comment]
@dsenalik Thanks, I will try your addition. Yea!

**Comment by srobb1** [comment]
I already have that line :(

`$base_table = $settings['base_table'];`

**Comment by srobb1** [comment]
@dsenalik if you try to add a new 'chado organism reference' to your gene page, do you get an error? This is a field that should already be in the drop down when you add a new field.

I am using the organism field as a template for the alternative name field, but I wanted to test that it actually worked before getting too far into my new field

**Comment by dsenalik** [comment]
@srobb1 That first line is just to show where the remaining lines go, it should be there and you add the remaining lines after (changing the cv term as appropriate) (I should have thought to do a diff🤷)
and yes, before the code change I got the same error as you did, after my fix it worked.

**Comment by dsenalik** [comment]
@srobb1 I made this into a pull request on the off chance that my change is correct.
@spficklin Can you check pull #1433 and pass your judgement of whether this is what we need to do for any field accessing the base_table? Or will it be obsolete if we change to yaml?

**Comment by srobb1** [comment]
As long as I have already created an organism, this works, though it is a bit weird that their needs to be a default organism.

**Comment by srobb1** [comment]
Also I was thinking it would be nice if the drop list of base tables was limited to only tables that have an organism_id fkey. I wonder how to do this. this would be good for any field, to limit to tables with a specific type of fk. Like the Chado Feature Sequence Field should be limited to base tables with feature_id .... well, it should be limited to only the feature table, right? There are no joins in that field, so I think it would only work with a base table of feature, so it would be good to just say the base table has to be feature for that one. Sorry for the rambling, just trying to make sense of this all.

**Comment by dsenalik** [comment]
Another problem with this field is that it doesn't work on ```project``` content types. Error:

```Warning: Undefined array key "foreign keys" in Drupal\tripal_chado\Plugin\Field\FieldType\obi__organism::tripalTypes() (line 89 of /var/www/drupal9/web/modules/contrib/tripal/tripal_chado/src/Plugin/Field/FieldType/obi__organism.php) ```

because for some reason there are no foreign keys defined for project. Here is the relevant code:
```
$base_schema_def = $schema->getTableDef($base_table, ['format' => 'Drupal']);
...
$base_fk_col = array_keys($base_schema_def['foreign keys']['organism']['columns'])[0];  // line causing error
```

You can confirm this in the devel module php window and see if the dumped variable contains a ```'foreign keys'``` array element. I tried just a few: feature, organism have ```'foreign keys'```, project, analysis do not:
```
$base_table = 'project';

$chado = \Drupal::service('tripal_chado.database');
$schema = $chado->schema();
$base_schema_def = $schema->getTableDef($base_table, ['format' => 'Drupal']);

var_dump($base_schema_def);
```

I have made this a separate issue #1436

**Comment by dsenalik** [comment]
Well this field is not supported on project because there is no ```organism_id``` in the project table. To show organisms for projects we would have to have a ```chado_linker__organism``` field that uses the ```project_organism``` table. And that table would have to be a custom table, as it does not exist. 🙈

**Comment by dsenalik** [comment]
This problem of being able to add a field where it is not supported is now issue #1440, I will remove the change I added for this from pull #1433 so that this issue is for only one thing.

**Comment by dsenalik** [comment]
So the remaining issue should be solved with merged PR #1433 so I am closing this issue.
